hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java

```diff
@@ -916,7 +917,7 @@ void waitForAckedSeqno(long seqno) throws IOException {
     try (TraceScope ignored = dfsClient.getTracer().
         newScope("waitForAckedSeqno")) {
       LOG.debug("{} waiting for ack for: {}", this, seqno);
-      int dnodes = nodes != null ? nodes.length : 3;
+      int dnodes = nodes.length > 0 ? nodes.length : 3;
```

Comment:
I think there is still a risk of a race here. In the original problem after nodes.length evaulated to say 3, then nodes was set to null. Then the "IF" statement tried to call null.length, leading to the error.

In this changed code, you could evaluate nodes.length to get 3, and then nodes is swapped to be the new `EMPTY_DATANODES`, then in the IF statement its length will be zero, setting dnodes to zero rather than 3, which will probably trip up the `getDatanodeWriteTimeout` method it is passed too.

To make it safe, you probably have to:

```
int currentNodes = nodes.length;
int dnodes = currentNodes > 0 ? currentNoides : 3;
```

Reply:
The nodes are not thread-safe. So int currentNodes = nodes.length will maybe NPE.    
Use thread-safe data structures or a lock?

Reply:
The nodes are protected by volatile  and it will be  assigned by EMPTY_DATANODES.  So I think odes.length will not be NPE.

Reply:
volatile does not guarantee that. You'd need something similar to AtomicInteger.compareAndSet()